### ISSUE TITLE
Serialization problem causes a stateful service to get stuck in a restart loop during upgrade

### ISSUE THREAD

**Issue by aelij** [Azure/service-fabric-issues#9]
Here's what happened:
- We upgraded a Service Fabric application with a stateful service that had some data
- The new Data Contract for that data was incompatible (we normally manage this using the `DataMemberAttribute.Order` property, but this time we missed one.)
- The service failed to load:

```
The application requested process termination through System.Environment.FailFast(string message).
Message: type cannot be null.
Stack:
   at System.Environment.FailFast(System.String)
   at Microsoft.ServiceFabric.Replicator.ReplicationMetadata..ctor(System.Uri, System.Type, Byte[], Int64, Int64, Int32, Microsoft.ServiceFabric.Replicator.StateManagerApplyType)
   at Microsoft.ServiceFabric.Replicator.ReplicationMetadata.Deserialize(Microsoft.ServiceFabric.Replicator.InMemoryBinaryReader, Microsoft.ServiceFabric.Replicator.InMemoryBinaryReader, System.String, Int32)
   at Microsoft.ServiceFabric.Replicator.DynamicStateManager+<OnApplyOnLocalStateAsync>d__ef.MoveNext()
```
- At this stage, the **upgrade was stuck**. Even rolling it back manually using PowerShell commands didn't work (it transitioned to rolling back, and remained stuck there). The service kept trying to load every 5 minutes or so and failed.
- We tried:
  - Restarting the machines
  - Restarting Service Fabric's Host Service
  - Deleting the service and repeating the above
- Deleting the service worked on 2 out of 3 nodes:
  - The last node remained with a single replica of the service (listed under the node in SF Explorer with the status `Down`)
  - The service was no longer listed when using `Get-ServiceFabricService`
  - The service still kept trying to load
- We finally invoked `Remove-ServiceFabricReplica` with the replica ID and restarted the SF Host Service on that node, which removed all the errors and allowed the upgrade to complete

I should also mention that all of this took quite a bit of time. Some PS commands take a while to respond and restarting nodes can also be slow. Luckily this all happened in a test environment. The error message ("type cannot be null") could also be improved.

**Comment by stpatrick2016** [comment]
When it happened to me, I came out of the restart loop by instructing removal or service and 10 seconds later killing the service's exe from task manager. Was much faster than doing anything else.
Have to mention, this was for Service Fabric Anywhere (e.g. on-premises). I guess this won't work on Azure.

**Comment by aelij** [comment]
@stpatrick2016 (Hey, Philip :) The process kept killing itself a few seconds after starting, so killing it didn't have any effect for us. And yes, this was on Azure, but you can still do this by connecting to the VM.

**Comment by sondreb** [comment]
Attempted to run the Remove-ServiceFabricReplica, when a service had been rolled back and issues with schema, so all replicas are down and a delete operation have already been executed for the service. It won't fully delete the service, since it has to be started to delete, and it won't start up, since it fails serialization.

So got this:

Remove-ServiceFabricReplica : This operation cannot be performed in the current replica state.

Which is probably because the replicas are in state "Down".

Any ideas on how to delete a service, which won't start up so it can be deleted?

**Comment by masnider** [comment]
@sondreb - did you use the -ForceRemove parameter with that?
@aelij - Generally when this happens you would update the parameters of the rollback so that it forces the change into the cluster. But even then, if you've messed up the state serialization, you'll need to push through another upgrade and force it through in order to read and re-write the state.

**Comment by aelij** [comment]
@masnider How do you 'force it through' in such a case? In the situation I described, why would the Fabric keep trying to load a deleted service on just one of the nodes? (Note this issue was opened a while ago, and I haven't tried recreating it in newer versions of the runtime.)

**Comment by sondreb** [comment]
I was able to get the service replicas to start up, by manually copying the binaries of the service from new version, to old version. That made the replicas start up, and then within a few seconds, they deleted successfully.

The -ForceRemove on Remove-ServiceFabricService ended up with a timeout, attempted that before replacing the binaries.

I would not advice anyone replacing binaries such as I did, but at least it fixed it for me.

**Comment by masnider** [comment]
Replacing the binaries is definitely not the right way to go about things :) Consider instead upgrades where you set the upgrade health maps to specifically allow certain services to not be healthy, and also to turn the timers/health checks down so that the code upgrade rolls through the cluster. That said when doing so you have to be sure you don't make things worse :)

**Comment by masnider** [comment]
@aelij if the service was there and we need to close it to drop it, we need to gracefully do so in order to have the service clean up its state or any resources that it has claimed. If it is not going to do that, then there is the potential for those resources to leak. The parameters are [here][1]. In particular, look at the health policy maps (so that you can indicate that a particular service is expected to be unhealthy) and UpgradeReplicaSetCheckTimeout. These will allow the code change to roll through the cluster and replace the service which are down. Generally this really should be the solution rather than deleting things - once you delete things and they are stuck then force removing things is the only route, which is leaky. 

[1]: https://docs.microsoft.com/en-us/azure/service-fabric/service-fabric-application-upgrade-parameters

**Comment by aelij** [comment]
@masnider What do you mean when you say "leak" here? What kind of resources? Files? Ports? Why is force-removal leaky?

**Comment by masnider** [comment]
TLDR - just don't delete things unless you are actually needing to remove them. If you're going to delete things, make sure that path works so you don't have to force remove them. If you do force remove them, you've probably leaked something and you have to take remedial action to make sure your environment stays clean. 

We don't rightly know. Think of it this way: all sorts of APIs will never fire. If the service is currently just running, things like ChangeRole->None (drop state on disk) or even ChangeRole Primary->Secondary won't happen. The service doesn't see the Close or Abort APIs, the communication listeners don't see close or abort. Anything that would normally be freed up there won't be. That could be ports, files, all sorts of things. Some of those depending on what they are and where they are stashed _might_ get cleaned up if the whole named application instance's directory goes away (which only happens when the whole named application instance is deleted). But since we don't know all the things that could have been claimed we can never say for sure. 

If you're trying to change things, upgrades are always preferable to rip and replace. If plan to delete things in prod, you really need to do thorough local testing to make sure your service can shut down cleanly in all sorts of states. [FailBrick][1] can show you some so you can practice getting out of them, but real chaos and upgrade & rollback and delete testing of your own workloads is the only way. 

[1]: https://github.com/masnider/FailBrick

**Comment by aelij** [comment]
@masnider Thanks! Sometimes I wish I could download some of your brain into mine :) Didn't know about FailBrick, looks like a neat tool.

Just to make things clear - so `-ForceRemove` just tears down the process without waiting for stuff like `ChangeRoleAsync` completion? I'm not sure I'd be concerned with a leak in this case (our services don't access shared resources other than what SF provides). A process can always *crash*, what's the difference? I'm assuming resources managed by SF (e.g. port lease, cert ref count, etc.) would be freed even when the service is force-deleted, right?

I wholeheartedly agree we should avoid deleting services, but as you can see from the chain of events above, I felt it was the last resort. Plus it was a test cluster. Next time this happens, I'll start by changing the timeouts as you suggested. My no. 1 wish for SF is still better *live* diagnostics tools for these potentially stressful situations... :)

**Comment by masnider** [comment]
@aelij yeah that's right, but not the process, the individual replica - it gets destructed. Which is part of why it can be weird - the process stays up. Anything the underlying OS considered claimed by the process stays allocated (what is that? we don't know). Just at the basics SF doesn't know what the service claimed, what handles were leaked or what resources won't be reclaimed, so we don't consider it really safe. 

Feedback heard around better diagnostics in these types of situations. As we see more of them we're improving things.

**Comment by NickDarvey** [comment]
Hi @aelij, we're also in this position.
In our case, I can't identify which DataContract has changed. (Actually, we have two in total and neither have...).
I can reproduce it by performing the upgrade locally but can't get any more information out of Service Fabric in order to determine where things are going wrong.

How did you figure out what was causing it?

**Comment by aelij** [comment]
@NickDarvey If neither of the DCs have changed, I'm not sure this is the same scenario. Is it possible you've added a new reliable dictionary with a new type during the upgrade? This can also cause a similar problem.

I've figured it out by thinking about it and then using ILSpy to verify my suspicion... :) The error message SF produces is quite obscure, but it gave me a hint there was a problem with the types loaded by the dictionary.

**Comment by NickDarvey** [comment]
@aelij yes we did add a dictionary actually! Is there a common mistake we might have made there?

New deployments worked fine but upgrades caused us to get into the same state you mentioned at the top.

**Comment by aelij** [comment]
@NickDarvey Adding a dictionary can be problematic during an upgrade, because some replicas which do not have the new types the dictionary uses may attempt to load it.

The Fabric team advised me this should be done in a two-phase upgrade: The first upgrade would add the types, and the other would actually create & use the dictionary. The second one can be a config-only update, activating a "feature flag" that enables the dictionary.

**Comment by aelij** [comment]
@NickDarvey You can also avoid creating the dictionary until the upgrade completes (somehow).

**Comment by masnider** [comment]
Multiple upgrades is the recommended way to do this. Yes because you are serializing types those types have to be everywhere before you can use or reference them. So one upgrade to provide the new type and then another upgrade that actually uses it anywhere.

**Comment by aelij** [comment]
@masnider I still contend that the Fabric should be _safer by default_. In this case even a rollback won't help, and we'd be stuck with a service that's potentially down for a long period of time.

One way to tackle this could be a setting that disallows adding new collections while an upgrade is running (i.e. `StateManager.GetOrAdd` would throw if a collection doesn't exist).

**Comment by NickDarvey** [comment]
@aelij @masnider multiple upgrades worked!
It'd be helpful if the error messaging made this clear, though it does make sense now why it occurs now it's been explained.

Getting it into this borked state is probably the most significant problem though, it's the kind of thing I would (as a relatively new SF application developer) hope rollback handles for me.

**Comment by masnider** [comment]
Hi @NickDarvey & @aelij 
Thanks for your patience and glad that things are working now as you expect. We're looking into what we can do both to make this scenario easier as well as how to improve diagnostics and monitoring in case things do go wrong. I'm going to close this issue down for now (since your situation is resolved) but know that we're still investing on it and taking your feedback into consideration.

**Comment by aelij** [comment]
@masnider Before closing this issue, I'd like to make sure that the error message has been improved. "type cannot be null" is just not clear enough. I think at the very least the error should say which type (and possibly which reliable dictionary and partition) were involved (in an ETW event?). Currently our only way of getting this information is by examining a dump. It would also be helpful if the error text was less like an assertion and more like an explanation of what the problem is (e.g. "Unable to load a reliable collection type ...").
